### PR TITLE
initial data & setData for composite & class components

### PR DESCRIPTION
**Pull request by AWoelfel**
I want to be able to use `setData` and `mount/shallowmount` `data: () => {}` callbacks to modify data in my composite and class components.

This should handle: 
- #1879 
- #1058 
and maby #1851 as well

### CODE REVIEW

**Review by cexbrayat — COMMENTED** [review]
As I mentioned in #1879 , I don't think supporting `setData`/initial data on components written with Composition API really makes sense, as the `data` field no longer exists in these components.

But I don't think strongly about it. @lmiller1990 @freakzlike what do you think ?

The same thing can be achieved via `wrapper.vm.prop = value`

File: src/mount.ts [hunk 1]
```diff
@@ -516,6 +516,24 @@ export function mount(
...
+          for (const [k, v] of Object.entries(
+            providedData as { [key: string]: any }
+          )) {
+            this.$.setupState[k] = v
+          }
+        }
+      }
+    })
+
+    app.mixin(mixin)
```
Comment (freakzlike) [inline src/mount.ts:534]: Wouldn't this be called for every component? Not only the component which has been passed to `mount`

**Review by freakzlike — COMMENTED** [review]
I don't feel comfortable using `data`/`setData` for Composition API. Maybe using a different name/functions for this?

File: src/mount.ts (hunk 1 above)
Comment (AWoelfel) [reply src/mount.ts:534]: you are right, as i added this as a mixin it will be called for every component in this App.
I will have a look if there is another convenient way to set the data fields.

**Comment by lmiller1990** [comment]
I agree with @cexbrayat, `setData` is `data` as in the Options API `data` function. This will be confusing.

Is there a case where you can't test your component without a `setSetupData` function? One of the explicit goals of the Script Setup is [closed by default](https://github.com/vuejs/rfcs/blob/master/active-rfcs/0040-script-setup.md#exposing-components-public-interface). The way Vue recommends exposing those is via `defineExposes`:

> To explicitly expose properties in a <script setup> component, use the defineExpose compiler macro

So, if we support this, would it makes sense to call it `setExposes` or `setExposedData`? You'd need to call `defineExposes` on whatever variables you'd like to modify in your tests, though.

**Comment by lmiller1990** [comment]
Left a comment ^

Interested in this but would like to at least discuss a tiny bit more - any new feature is more complexity and maintenance so we should carefully consider based on use cases, edge cases, back/forward compatibility, etc.

**Comment by merkle-rking** [comment]
> Is there a case where you can't test your component without a `setSetupData` function? I haven't encountered a compelling use case so far - I usually just set the state how a user does, by triggering events, etc.

We've recently migrated from Vue 2 to Vue 3 and are embracing [knowing what to test](https://v1.test-utils.vuejs.org/guides/#common-tips) given the changes when moving to the Composition API. It's been quite a mentality switch from our 100% line coverage approach in Vue 2.

Today, I may have run into such a case where having access to the underlying data in the setup() would be helpful, if not necessary. I'm trying to write tests around a file input component. The file input's value cannot be directly modified for security reasons:

`InvalidStateError: This input element accepts a filename, which may only be programmatically set to the empty string.`

I'm having trouble figuring out how to get the input populated with a File() object in our tests so I can test various things (validation, the file being passed through to an API call upon form submission, etc.). Any thoughts on how to bridge this gap would be appreciated, but it may be the use-case you're looking for.

I should also mention that while we have been able to adapt to the idea that we have to emulate user actions to set values in controls, i.e. element.trigger('click'), element.setValue('foo'), etc, there's a noticeable friction in this testing when compared to the old methodology. As an example, Vuetify 3.1 recently made some changes to some of their controls and layouts that are really unimportant to us from an implementation standpoint. They changed some of the control activation events from "click" to "mousedown". These have the same effect in a browser and are almost entirely transparent to the user. For us developers, this means our tests failed upon upgrade and we had to refactor our emulated user actions. In the Options API, we would have just set the data directly with the expectation that our controls' internals are unimportant. Locating elements used in Teleports is another such example. Instead of finding an array of elements and clicking on all of them, we would've just set the data to an array in our tests.

Admittedly, the new approach gives us the stronger assertion that our controls are correctly tied to data elements because we're forced to interact with the control to engage the data.

**Comment by merkle-rking** [comment]
Coincidentally, I may have stumbled on a second use-case. We have a component that allows for updating a resource server-side. When that resource is updated, the change is logged. On the front-end, we have a View that renders both that resource and the log of the changes to the resource. When the resource is changed, we also want to have the log events on page get refreshed. The log events rendering is abstracted into its own separate component. To facilitate the forced re-render, we're using they key-changing technique. Within our tests for the main View, the log component is stubbed. The testing for the log component is separated. Our snapshots in the main View contain the stubbed representation of our log component, but because there is no representation of the `:key` in the stub, we're unable to test that when a user does an action in the main View, it forces a re-render of the log component. One alternative solution to exposing setup() data would be to have `:key` be reflected in the stub so we can assert against the value incrementing in our snapshots when we initiate user clicks. We could of course mount the log component directly in the main View's tests, but it's much cleaner to separate the concerns.

**Comment by cexbrayat** [comment]
Superseeded by https://github.com/vuejs/test-utils/pull/2166
